### PR TITLE
Tarred Datasets for Monolingual Corpora

### PR DESCRIPTION
**Pull request by MaximumEntropy**
This PR introduces tarred datasets for monolingual text corpora to allow (back)translating large amounts of text. A tarred dataset can be created as follows:

```
python examples/nlp/machine_translation/create_tarred_monolingual_dataset.py \
    --tokenizer_model /path/to/tokenizer.model \
    --clean \
    --fname /path/to/monolingual_data.txt \
    --max_seq_length 128 \
    --min_seq_length 1 \
    --tokens_in_batch 32000 \
    --lines_per_dataset_fragment 5000000 \
    --num_batches_per_tarfile 500 \
    --out_dir /path/to/output_folder
```

Once the `output_folder` is created, we can batch translate data using a pre-trained model

```
python examples/nlp/machine_translation/translate_ddp.py \
    --model /path/to/pretrained_model.nemo \
    --text2translate /path/to/output_folder/batches.tokens.32000._OP_1..100_CL_.tar \
    --metadata_path /path/to/output_folder/metadata.json \
    --result_dir /path/to/translations \
    --topk 500 \
    --src_language en \
    --tgt_language de \
    --n_gpus 8
```

### CODE REVIEW

File: examples/nlp/machine_translation/create_tarred_parallel_dataset.py [hunk 1]
```diff
@@ -111,6 +112,7 @@ def write_batches_to_tarfiles(
             data='/tmp/concat_dataset.txt',
             vocab_size=args.vocab_size,
             model=os.path.join(args.out_dir, 'tokenizer.%d.BPE.model' % (args.vocab_size)),
+            coverage=args.coverage,
```
Comment (okuchaiev) [inline examples/nlp/machine_translation/create_tarred_parallel_dataset.py:115]: I am wondering if this should be supported by automatic tarred dataset creation @ericharper ?

**Review by okuchaiev — COMMENTED** [review]
@MaximumEntropy  and @ericharper should we refactor ``examples/nlp/machine_translation/create_tarred_parallel_dataset.py`` to rely on logic in ``nemo.collections.nlp.data.machine_translation.preproc_mt_data.MTDataPreproc`` ?

File: examples/nlp/machine_translation/create_tarred_parallel_dataset.py (hunk 1 above)
Comment (MaximumEntropy) [reply examples/nlp/machine_translation/create_tarred_parallel_dataset.py:115]: Happy to include it in `preproc_mt_data.py`

Comment (ericharper) [reply examples/nlp/machine_translation/create_tarred_parallel_dataset.py:115]: That would be great!

**Comment by ericharper** [comment]
> @MaximumEntropy and @ericharper should we refactor `examples/nlp/machine_translation/create_tarred_parallel_dataset.py` to rely on logic in `nemo.collections.nlp.data.machine_translation.preproc_mt_data.MTDataPreproc` ?

Maybe we can just use the MTDataPreproc since we have the do_training=False flag?

File: examples/nlp/machine_translation/create_tarred_monolingual_dataset.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            )
+            num_files_in_tar = 0
+
+        tar_file_ptr.add(os.path.join(args.out_dir, 'batch-%d.pkl' % (global_batch_ctr)))
+        num_files_in_tar += 1
+        os.remove(os.path.join(args.out_dir, 'batch-%d.pkl' % (global_batch_ctr)))
+    return tar_file_ptr, global_batch_ctr, num_files_in_tar, tar_file_ctr
+
+
+if __name__ == '__main__':
```
Comment (okuchaiev) [inline examples/nlp/machine_translation/create_tarred_monolingual_dataset.py:65]: Can you please refactor this and move this logic to ``nemo.collections.nlp.data.machine_translation.preproc_mt_data.MTDataPreproc`` and just call a method of nemo.collections.nlp.data.machine_translation.preproc_mt_data.MTDataPreproc in this script?

**Review by okuchaiev — CHANGES_REQUESTED** [review]

**Comment by okuchaiev** [comment]
> > @MaximumEntropy and @ericharper should we refactor `examples/nlp/machine_translation/create_tarred_parallel_dataset.py` to rely on logic in `nemo.collections.nlp.data.machine_translation.preproc_mt_data.MTDataPreproc` ?
> 
> Maybe we can just use the MTDataPreproc since we have the do_training=False flag?

I think it is fine and useful to have these two scripts, my biggest concern is that the actual logic needs to be implemented in one place (MTDataPreproc class probably)

File: examples/nlp/machine_translation/translate_ddp.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                    o = sp_detokenizer.detokenize(o.split())
+                translation = source_detokenizer.detokenize(o.split())
+                of.write(translation + '\n')
+    cleanup()
+
+
+def main() -> None:
+    world_size = torch.cuda.device_count()
+    args = get_args()
+    mp.spawn(translate, args=(world_size, args), nprocs=world_size, join=True)
```
Comment (ericharper) [inline examples/nlp/machine_translation/translate_ddp.py:160]: I'm curious why you didn't use PyTorch Lightning `Trainer.test` for this? The idea behind using PTL is minimize having to write this kind of boiler plate code.

File: nemo/collections/nlp/data/machine_translation/one_side_dataset.py [hunk 4]
```diff
@@ -131,3 +136,137 @@ def clean(self, ids, max_tokens=None, min_tokens=None):
...
+        self.tarpath = text_tar_filepaths
+
+        # Put together WebDataset
+        self._dataset = (
+            wd.Dataset(text_tar_filepaths)
+            .shuffle(shuffle_n)
+            .rename(pkl='pkl', key='__key__')
+            .to_tuple('pkl', 'key')
+            .map(f=self._build_sample)
+        )
```
Comment (ericharper) [inline nemo/collections/nlp/data/machine_translation/one_side_dataset.py:256]: I believe the syntax has changed for WebDataset with the recent version upgrade: https://github.com/NVIDIA/NeMo/pull/1760
https://github.com/NVIDIA/NeMo/pull/1760/files#diff-26041722ee3bb4aa3830321d28b90ed90eefcc10ba55292d9a7a74e0b4216281R400
Please run ./reinstall.sh or updgrade Webdataset.

File: examples/nlp/machine_translation/translate_ddp.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    parser.add_argument("--result_dir", type=str, required=True, help="Folder to write translation results")
+    parser.add_argument(
+        "--twoside", action="store_true", help="Set flag when translating the source side of a parallel dataset"
+    )
+    parser.add_argument(
+        '--metadata_path', type=str, required=True, help="Path to the JSON file that contains dataset info"
+    )
+    parser.add_argument('--topk', type=int, default=500, help="Value of k for topk sampling")
+    parser.add_argument('--source_lang', type=str, required=True, help="Source lang ID for detokenization")
+    parser.add_argument('--target_lang', type=str, required=True, help="Target lang ID for detokenization")
```
Comment (ericharper) [inline examples/nlp/machine_translation/translate_ddp.py:48]: can we use `src_language` and `tgt_language` to match the model config?

File: examples/nlp/machine_translation/translate_ddp.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+from nemo.collections.common.tokenizers.pangu_jieba_detokenizer import PanguJiebaDetokenizer
+from nemo.collections.common.tokenizers.sentencepiece_detokenizer import SentencePieceDetokenizer
+from nemo.collections.nlp.data.machine_translation import TarredOneSideTranslationDataset, TarredTranslationDataset
+from nemo.collections.nlp.models.machine_translation.mt_enc_dec_model import MTEncDecModel
+from nemo.utils import logging
+
+
+def get_args():
+    parser = ArgumentParser(description='Batch translation of sentences from a pre-trained model on multiple GPUs')
```
Comment (ericharper) [inline examples/nlp/machine_translation/translate_ddp.py:34]: We should have an argument to specify number of GPUs to be used.

File: examples/nlp/machine_translation/translate_ddp.py (hunk 3 above)
Comment (ericharper) [reply examples/nlp/machine_translation/translate_ddp.py:160]: Also, I'm actually having trouble getting this to run on my system with mixed GPUs. One of my GPUs is dedicated to graphics, but this script keeps trying to use it.

File: nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py [hunk 7]
```diff
@@ -164,6 +164,10 @@ def filter_predicted_ids(self, ids):
     @typecheck()
     def forward(self, src, src_mask, tgt, tgt_mask):
         src_hiddens = self.encoder(src, src_mask)
+        if tgt is None:
+            beam_results = self.beam_search(encoder_hidden_states=src_hiddens, encoder_input_mask=src_mask)
+            beam_results = self.filter_predicted_ids(beam_results)
+            return beam_results
```
Comment (ericharper) [inline nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py:170]: We just took beam search out of the forward pass.

**Review by okuchaiev — APPROVED** [review]
looks good to me now. please get @ericharper 's approval

File: examples/nlp/machine_translation/translate_ddp.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+    # initialize the process group
+    dist.init_process_group("gloo", rank=rank, world_size=world_size)
+
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+def translate(rank, world_size, args):
```
Comment (ericharper) [inline examples/nlp/machine_translation/translate_ddp.py:70]: Let's add a `batch_translate` method so logic can all be in one place.

**Review by ericharper — CHANGES_REQUESTED** [review]
Really nice PR. Please address the comments then we can approve.

**Review by ericharper — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**